File: sdks/java-agent/src/main/java/com/carhire/sdk/TransportException.java
package com.carhire.sdk;

public class TransportException extends Exception {
    private final Integer statusCode;
    private final String code;

    public TransportException(String message) {
        super(message);
        this.statusCode = null;
        this.code = null;
    }

    public TransportException(String message, Integer statusCode, String code) {
        super(message);
        this.statusCode = statusCode;
        this.code = code;
    }

    public Integer getStatusCode() {
        return statusCode;
    }

    public String getCode() {
        return code;
    }

    public static TransportException fromHttp(Exception error) {
        // Extract status code and message from HTTP error
        String message = error.getMessage();
        Integer statusCode = null;
        String code = null;

        if (error instanceof okhttp3.HttpException) {
            okhttp3.HttpException httpError = (okhttp3.HttpException) error;
            statusCode = httpError.code();
            message = httpError.message();
        }

        return new TransportException(message, statusCode, code);
    }

    public static TransportException fromGrpc(Exception error) {
        String message = error.getMessage();
        String code = null;

        if (error instanceof io.grpc.StatusException) {
            io.grpc.StatusException grpcError = (io.grpc.StatusException) error;
            code = grpcError.getStatus().getCode().toString();
            message = grpcError.getStatus().getDescription();
        }

        return new TransportException(message, null, code);
    }
}

File: sdks/nodejs-agent/src/Exceptions/TransportException.ts
export class TransportException extends Error {
  public readonly statusCode?: number;
  public readonly code?: string;

  constructor(message: string, statusCode?: number, code?: string) {
    super(message);
    this.name = 'TransportException';
    this.statusCode = statusCode;
    this.code = code;
    Error.captureStackTrace(this, this.constructor);
  }

  public static fromHttp(error: unknown): TransportException {
    if (error instanceof Error) {
      const axiosError = error as { response?: { status?: number; data?: unknown }; code?: string };
      const statusCode = axiosError.response?.status;
      const message = axiosError.response?.data
        ? JSON.stringify(axiosError.response.data)
        : error.message;
      return new TransportException(message, statusCode, axiosError.code);
    }
    return new TransportException(String(error));
  }

  public static fromGrpc(error: unknown): TransportException {
    if (error instanceof Error) {
      const grpcError = error as { code?: number; details?: string };
      return new TransportException(
        grpcError.details || error.message,
        grpcError.code,
        String(grpcError.code)
      );
    }
    return new TransportException(String(error));
  }
}

